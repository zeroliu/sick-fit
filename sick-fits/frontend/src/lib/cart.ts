import { CartItem } from 'src/queries/user';

export function calcTotalPrice(cartItems: CartItem[]) {
  return cartItems.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity * cartItem.item.price;
  }, 0);
}

export function countItems(cartItems: CartItem[]) {
  return cartItems.reduce((tally, cartItem) => {
    if (!cartItem.item) return tally;
    return tally + cartItem.quantity;
  }, 0);
}
