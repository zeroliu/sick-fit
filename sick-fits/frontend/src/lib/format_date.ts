export function formatDate(dateStr: string) {
  const date = new Date(dateStr);
  const options = {
    year: 'numeric',
    month: 'numeric',
    day: 'numeric',
    hour: 'numeric',
    minute: 'numeric',
    second: 'numeric',
    timeZoneName: 'short',
  };
  return new Intl.DateTimeFormat('en-US', options).format(date);
}
