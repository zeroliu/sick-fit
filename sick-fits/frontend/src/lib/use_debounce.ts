import { useRef, useCallback, useEffect } from 'react';

export function useDebounce<Args extends unknown[]>(
  callback: (...args: Args) => void,
  timeout: number,
): (...args: Args) => void {
  const tid = useRef(-1);
  const debouncedFunc = useRef(callback);
  const hasUnmounted = useRef(false);
  debouncedFunc.current = callback;

  useEffect(
    () => () => {
      hasUnmounted.current = true;
    },
    [],
  );

  const debouncedCallback = useCallback(
    (...args: Args) => {
      clearTimeout(tid.current);
      tid.current = setTimeout(() => {
        if (hasUnmounted.current) {
          return;
        }
        debouncedFunc.current(...args);
      }, timeout);
    },
    [timeout],
  );

  return debouncedCallback;
}
