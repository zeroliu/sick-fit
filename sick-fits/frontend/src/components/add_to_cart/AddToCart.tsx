import React from 'react';
import { useAddToCartMutation } from 'src/queries/cart';
import { ME_QUERY } from 'src/queries/user';

interface Props {
  id: string;
}
export const AddToCart: React.FC<Props> = ({ id: itemId }) => {
  const [addToCart, { error }] = useAddToCartMutation({
    variables: {
      data: {
        itemId,
      },
    },
    refetchQueries: [{ query: ME_QUERY }],
  });
  if (error) {
    alert(error.message);
  }

  return <button onClick={() => addToCart()}>Add To Cart</button>;
};
