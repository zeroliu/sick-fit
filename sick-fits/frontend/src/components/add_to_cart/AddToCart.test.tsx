import { MockedProvider } from '@apollo/react-testing';
import { render, waitFor } from '@testing-library/react';
import React from 'react';

import { AddToCart } from './AddToCart';
import { createSuccessMocks, createFailureMocks } from './fixtures/default';
import { initStoryshots } from 'src/lib/test_utils';

initStoryshots();

describe('<AddToCart>', () => {
  it('triggers ADD_TO_CART mutation when clicked', async () => {
    const itemId = '1';
    const triggerAddToCart = jest.fn();
    const { getByTestId } = render(
      <MockedProvider
        mocks={createSuccessMocks(itemId, {
          addToCartMutationCallback: triggerAddToCart,
        })}>
        <AddToCart id={itemId} />
      </MockedProvider>,
    );
    getByTestId('button').click();
    await waitFor(() => expect(triggerAddToCart).toHaveBeenCalled());
  });

  it('refetches ME_QUERY', async () => {
    const itemId = '1';
    const triggerMeQuery = jest.fn();
    const { getByTestId } = render(
      <MockedProvider
        mocks={createSuccessMocks(itemId, {
          meQueryCallback: triggerMeQuery,
        })}>
        <AddToCart id={itemId} />
      </MockedProvider>,
    );
    getByTestId('button').click();
    await waitFor(() => expect(triggerMeQuery).toHaveBeenCalled());
  });

  it('alerts the error when failed', async () => {
    const itemId = '1';
    const alertSpy = jest.spyOn(window, 'alert').mockImplementation(() => {
      // noop.
    });
    const { getByTestId } = render(
      <MockedProvider mocks={createFailureMocks(itemId)}>
        <AddToCart id={itemId} />
      </MockedProvider>,
    );
    getByTestId('button').click();
    await waitFor(() =>
      expect(alertSpy).toHaveBeenCalledWith('Network error: it failed'),
    );
  });
});
