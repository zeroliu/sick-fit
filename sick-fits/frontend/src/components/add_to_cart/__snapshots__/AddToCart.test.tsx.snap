// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storyshots <AddToCart> Base 1`] = `
<button
  data-testid="button"
>
  Add To Cart
</button>
`;

exports[`Storyshots <CartCount> Count 5 1`] = `
<span
  class="cart_count_styles__AnimatedCartCount-sc-18ipxui-1 iyarbG"
>
  <div
    class="cart_count_styles__Dot-sc-18ipxui-0 fPUWhS count"
  >
    5
  </div>
</span>
`;

exports[`Storyshots <CartCount> Count 10 1`] = `
<span
  class="cart_count_styles__AnimatedCartCount-sc-18ipxui-1 iyarbG"
>
  <div
    class="cart_count_styles__Dot-sc-18ipxui-0 fPUWhS count"
  >
    10
  </div>
</span>
`;
